File: sdks/aas-core3-typescript/emit_report.js
#!/usr/bin/env node
/**
 * Convert tinybench JSON output to report.json schema.
 *
 * Usage:
 *   node emit_report.js <bench_raw.json> <output_path>
 *
 * tinybench results have `mean` in milliseconds — we convert to nanoseconds.
 */

import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const CORE_DATASETS = new Set(["wide", "deep", "mixed"]);
const CORE_OPERATIONS = new Set(["deserialize", "validate", "traverse", "update", "serialize"]);
const XML_OPERATIONS = new Set(["deserialize_xml", "serialize_xml"]);
const AASX_OPERATIONS = new Set(["aasx_extract", "aasx_repackage"]);

function msToNs(ms) {
  if (ms === null || ms === undefined) {
    return null;
  }
  return Math.round(ms * 1e6);
}

function normalizeOperationId(operation) {
  const lower = operation.toLowerCase();
  const map = {
    deserializexml: "deserialize_xml",
    serializexml: "serialize_xml",
    aasxextract: "aasx_extract",
    aasxrepackage: "aasx_repackage",
  };
  if (map[lower]) return map[lower];
  return operation;
}

function inferOperationTrack(dataset, operationId) {
  if (XML_OPERATIONS.has(operationId)) return "xml";
  if (AASX_OPERATIONS.has(operationId)) return "aasx";
  if (dataset.startsWith("val_") && operationId === "validate") return "validation";
  if (CORE_DATASETS.has(dataset) && CORE_OPERATIONS.has(operationId)) return "core";
  return "capability";
}

function main() {
  if (process.argv.length !== 4) {
    process.stderr.write(
      "Usage: node emit_report.js <bench_raw.json> <output_path>\n"
    );
    process.exit(1);
  }

  const inputPath = process.argv[2];
  const outputPath = process.argv[3];

  const rawData = fs.readFileSync(inputPath, "utf-8");
  const benchResults = JSON.parse(rawData);

  // Organize by dataset
  const datasets = {};

  for (const result of benchResults) {
    const dsName = result.dataset;
    if (!(dsName in datasets)) {
      datasets[dsName] = {
        file_size_bytes: null,
        element_count: null,
        operations: {},
      };
    }

    const meanNs = msToNs(result.meanMs);
    let throughput = 0;
    if (result.meanMs > 0) {
      throughput = Math.round((1000 / result.meanMs) * 100) / 100;
    }

    const memoryEntry = {
      peak_rss_bytes: null,
      heap_used_bytes: null,
      alloc_bytes_per_op: null,
      alloc_count_per_op: null,
      gc_pause_ms: null,
      gc_count: null,
      traced_peak_bytes: null,
    };

    if (result.memory) {
      memoryEntry.peak_rss_bytes =
        result.memory.peakRssBytes > 0 ? result.memory.peakRssBytes : null;
      memoryEntry.heap_used_bytes =
        result.memory.heapUsedBytes != null && result.memory.heapUsedBytes !== 0
          ? result.memory.heapUsedBytes
          : null;
    }

    const operationId = normalizeOperationId(result.operation);
    datasets[dsName].operations[operationId] = {
      operation_id: operationId,
      operation_track: inferOperationTrack(dsName, operationId),
      sample_count: result.sampleCount || 0,
      measurement_semantics: "mean_ns_per_operation",
      failure_state: "ok",
      iterations: result.iterations || 0,
      mean_ns: meanNs,
      median_ns: msToNs(result.medianMs),
      stddev_ns: msToNs(result.stddevMs),
      min_ns: msToNs(result.minMs),
      max_ns: msToNs(result.maxMs),
      p75_ns: msToNs(result.p75Ms),
      p99_ns: msToNs(result.p99Ms),
      throughput_ops_per_sec: throughput,
      memory: memoryEntry,
    };
  }

  // Detect Node.js version
  const nodeVersion = process.version;

  // Try to get SDK package version
  let sdkVersion = "unknown";
  try {
    const pkgPath = path.join(
      __dirname,
      "node_modules",
      "@aas-core-works",
      "aas-core3.0-typescript",
      "package.json"
    );
    if (fs.existsSync(pkgPath)) {
      const pkg = JSON.parse(fs.readFileSync(pkgPath, "utf-8"));
      sdkVersion = pkg.version || "unknown";
    }
  } catch {
    // ignore
  }

  const report = {
    schema_version: 2,
    sdk_id: "aas-core3-typescript",
    metadata: {
      language: "typescript",
      runtime_version: `Node.js ${nodeVersion}`,
      sdk_package_version: sdkVersion,
      benchmark_harness: "tinybench",
      timestamp: new Date().toISOString().replace(/\.\d{3}Z$/, "Z"),
    },
    datasets,
  };

  // Ensure output directory exists
  const outDir = path.dirname(outputPath);
  if (!fs.existsSync(outDir)) {
    fs.mkdirSync(outDir, { recursive: true });
  }

  fs.writeFileSync(outputPath, JSON.stringify(report, null, 2) + "\n", "utf-8");
  process.stderr.write(`Wrote report to ${outputPath}\n`);
}

main();

File: sdks/basyx-rust/emit_report.py
#!/usr/bin/env python3
"""Convert criterion benchmark results to report.json schema v2.

Usage:
    python3 emit_report.py <criterion_target_dir> <output_path>

Criterion stores results in target/criterion/<group>/<benchmark>/new/estimates.json
with times in nanoseconds.

Schema v2 additions:
  - memory.heap_used_bytes, gc_pause_ms, gc_count, traced_peak_bytes
  - peak_rss_bytes read from /proc/self/status VmHWM (Linux CI only)
"""
import json
import os
import sys
from datetime import datetime, timezone
from pathlib import Path

CORE_DATASETS = {"wide", "deep", "mixed"}
CORE_OPERATIONS = {"deserialize", "validate", "traverse", "update", "serialize"}


def infer_operation_track(dataset: str, operation_id: str) -> str:
    """Classify operation into the core track or capability tracks."""
    if operation_id in {"deserialize_xml", "serialize_xml"}:
        return "xml"
    if operation_id in {"aasx_extract", "aasx_repackage"}:
        return "aasx"
    if dataset.startswith("val_") and operation_id == "validate":
        return "validation"
    if dataset in CORE_DATASETS and operation_id in CORE_OPERATIONS:
        return "core"
    return "capability"


def _read_vmhwm_bytes():
    """Read VmHWM (peak resident set size) from /proc/self/status.

    Returns the value in bytes, or None if unavailable (non-Linux).
    """
    try:
        status_path = Path("/proc/self/status")
        if not status_path.exists():
            return None
        for line in status_path.read_text().splitlines():
            if line.startswith("VmHWM:"):
                # Format: "VmHWM:    123456 kB"
                parts = line.split()
                if len(parts) >= 2:
                    kb_value = int(parts[1])
                    return kb_value * 1024
    except Exception:
        pass
    return None


def main():
    if len(sys.argv) != 3:
        print("Usage: python3 emit_report.py <criterion_dir> <output_path>", file=sys.stderr)
        sys.exit(1)

    criterion_dir = Path(sys.argv[1])
    output_path = sys.argv[2]

    # Read peak RSS early (will be None on non-Linux)
    peak_rss_bytes = _read_vmhwm_bytes()

    datasets = {}

    # Criterion directory structure: <group>/<benchmark>/new/estimates.json
    # group = operation (deserialize, serialize)
    # benchmark = dataset name (wide, deep, mixed)
    if criterion_dir.exists():
        for group_dir in sorted(criterion_dir.iterdir()):
            if not group_dir.is_dir():
                continue
            operation = group_dir.name
            if operation.startswith("report"):
                continue

            for bench_dir in sorted(group_dir.iterdir()):
                if not bench_dir.is_dir():
                    continue
                ds_name = bench_dir.name

                estimates_path = bench_dir / "new" / "estimates.json"
                if not estimates_path.exists():
                    continue

                with open(estimates_path) as f:
                    estimates = json.load(f)

                mean_ns = round(estimates.get("mean", {}).get("point_estimate", 0))
                median_ns = round(estimates.get("median", {}).get("point_estimate", 0))
                stddev_ns = round(estimates.get("std_dev", {}).get("point_estimate", 0))

                # Load sample data if available
                sample_path = bench_dir / "new" / "sample.json"
                min_ns = None
                max_ns = None
                iterations = 0
                sample_count = 0
                if sample_path.exists():
                    with open(sample_path) as f:
                        sample = json.load(f)
                    times = sample.get("times", [])
                    iters = sample.get("iters", [])
                    sample_count = len(times)
                    if times and iters:
                        per_op = [t / i for t, i in zip(times, iters) if i > 0]
                        if per_op:
                            min_ns = round(min(per_op))
                            max_ns = round(max(per_op))
                        iterations = sum(int(i) for i in iters)

                throughput = 0
                if mean_ns > 0:
                    throughput = round(1e9 / mean_ns, 2)

                if ds_name not in datasets:
                    datasets[ds_name] = {
                        "file_size_bytes": None,
                        "element_count": None,
                        "operations": {},
                    }

                datasets[ds_name]["operations"][operation] = {
                    "operation_id": operation,
                    "operation_track": infer_operation_track(ds_name, operation),
                    "sample_count": sample_count,
                    "measurement_semantics": "mean_ns_per_operation",
                    "failure_state": "ok",
                    "iterations": iterations,
                    "mean_ns": mean_ns,
                    "median_ns": median_ns,
                    "stddev_ns": stddev_ns,
                    "min_ns": min_ns,
                    "max_ns": max_ns,
                    "p75_ns": None,
                    "p99_ns": None,
                    "throughput_ops_per_sec": throughput,
                    "memory": {
                        "peak_rss_bytes": peak_rss_bytes,
                        "alloc_bytes_per_op": None,
                        "alloc_count_per_op": None,
                        "heap_used_bytes": None,
                        "gc_pause_ms": None,
                        "gc_count": None,
                        "traced_peak_bytes": None,
                    },
                }

    # Detect Rust version
    rust_version = "unknown"
    try:
        import subprocess
        result = subprocess.run(["rustc", "--version"], capture_output=True, text=True)
        rust_version = result.stdout.strip()
    except Exception:
        pass

    # Get crate version from Cargo.toml
    sdk_version = "unknown"
    cargo_path = Path(__file__).parent / "Cargo.toml"
    if cargo_path.exists():
        for line in cargo_path.read_text().splitlines():
            if "basyx-rs" in line and "=" in line:
                sdk_version = line.split("=")[-1].strip().strip('"')
                break

    report = {
        "schema_version": 2,
        "sdk_id": "basyx-rust",
        "metadata": {
            "language": "rust",
            "runtime_version": rust_version,
            "sdk_package_version": sdk_version,
            "benchmark_harness": "criterion",
            "timestamp": datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ"),
        },
        "datasets": datasets,
    }

    out_dir = Path(output_path).parent
    out_dir.mkdir(parents=True, exist_ok=True)

    with open(output_path, "w") as f:
        json.dump(report, f, indent=2)
        f.write("\n")

    print(f"Wrote report to {output_path}", file=sys.stderr)
    if peak_rss_bytes is not None:
        print(f"  peak_rss_bytes (VmHWM): {peak_rss_bytes}", file=sys.stderr)
    else:
        print("  peak_rss_bytes: unavailable (non-Linux or /proc not mounted)", file=sys.stderr)


if __name__ == "__main__":
    main()
